pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-spark-3/src/main/java/org/apache/pinot/plugin/ingestion/batch/spark3/SparkSegmentMetadataPushJobRunner.java

```diff
@@ -175,28 +192,52 @@ private void handleNonConsistentPush(List<String> segmentsToPush, PinotFS output
     } else {
       JavaSparkContext sparkContext = JavaSparkContext.fromSparkContext(SparkContext.getOrCreate());
       JavaRDD<String> pathRDD = sparkContext.parallelize(segmentsToPush, pushParallelism);
-      // Prevent using lambda expression in Spark to avoid potential serialization exceptions, use inner function
-      // instead.
-      pathRDD.foreach(new VoidFunction<String>() {
-        @Override
-        public void call(String segmentTarPath)
-            throws Exception {
-          PluginManager.get().init();
-          setupFileSystems();
-          try {
-            Map<String, String> segmentUriToTarPathMap =
-                SegmentPushUtils.getSegmentUriToTarPathMap(outputDirURI, _spec.getPushJobSpec(),
-                    new String[]{segmentTarPath});
-            SegmentPushUtils.sendSegmentUriAndMetadata(_spec, PinotFSFactory.create(outputDirURI.getScheme()),
-                segmentUriToTarPathMap);
-          } catch (RetriableOperationException | AttemptsExceededException e) {
-            throw new RuntimeException(e);
+
+      if (_spec.getPushJobSpec().isBatchSegmentUpload()) {
+        // Process segments in batch mode using foreachPartition
+        pathRDD.foreachPartition(new VoidFunction<Iterator<String>>() {
```

Comment:
while sendSegmentUriAndMetadata can accept multiple segments in its segmentUriToTarPathMap, and batch the call, it defeats the purpose of the spark executors.

so this implementation here ensures that batching happens at the executor level, and not the driver (unless you do push parallelism == 1 AND batch upload = true). 

let's assume push parallelism == 4 and segments == 40 and batch upload = true.
We have have 4 spark tasks, each task will have 10 segments, and each executor will submit _10_ segments as part of the batch upload.

let's assume push parallelism == 1 and segments == 40 and batch upload = true.
We will have 1 spark task, with 40 segments that will be submitted as part of the batch upload.

let's assume push parallellism == 4, segments = 40, batch upload = false
We will have 40 spark tasks, 10 assigned to each executor, who will upload 1 by 1 but in parallel (so 4 executors working at the same time).

we should note all these options in the pinot docs so users understand how batch mode interplays with parallelism

Reply:
Agree on the documentation part. 

Yes batch upload is a new optimization, which was very useful for StarTree usecase to eliminate the contention/bottleneck during ideal state update when segments are added one at a time.
In this case its very relevant because we've already got all segments to upload.  To start with, if batch mode is enabled, we dont need push parallelism to also be present? With a single upload (all or nothing), we should be able to upload all segments, that are in the same consistent push session
cc @rajagopr

Reply:
Theoretically we shouldn't _need_ push paralllelism if batch mode is true, but in the current implementation, its used to further sub-dive the tasks (ie, 40 segments, with 4 push parlalleism w/ batch upload == 10 segments/batch upload). I imagine at large scales (50k+ segments) this could be useful in the event that 50k segments in 1 batch upload doesnt work but say 10k/batch does work.

If we limit it so that push parallellism must be 1 when batch upload is true, we may be limiting our options. Though w/o emperically testing what is the limit of the batch upload API, we dont know for sure. 

let's see if @rajagopr has any suggestion or info here

Reply:
Yes we can limit the size of each batch with pushParallelism option, when the payload per push is large. Given that we compress the metadata, I would think that this api would be able to support large number of segments (1000's).

Reply:
myself/[swaminathanmanish](https://github.com/swaminathanmanish)/@rajagopr discussed this in slack and we are happy with the current implementation in this PR since it gives users the flexibility to do it all in 1 go (push parallelism == 1) or increase it if the scale is too large to do in 1 go.